Swep el patito e.r.c.s

ELPatitoLoco opened the pull request:
## Changes

I added a new swep and changed tas_weapons_init.lua to add myself as a Author hope it doesn't create conflicts, for everything else I think my changes won't affect anything

I left a commented part at the bottom of my code, as it says on the comment for the moment it doesn't work and I don't wanna cause conflicts due to it, in the future I'll make updates to my swep.

## Helpful Links

[Discord](https://discord.tasevers.com)

ELPatitoLoco committed:
V1.0 Just uploading the basic coede

ELPatitoLoco committed:
Uploading code

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.CannonForce = 50000000
+
+function SWEP:Initialize()
+	self:SetWeaponHoldType("revolver")
```
nit: Use `Weapon:SetHoldType(string name )` instead.

ELPatitoLoco replied:
Done

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+	Bullet.Tracer = self.Primary.Tracer
+	Bullet.Damage = self.Primary.Damage
+	Bullet.AmmoType = self.Primary.Ammo
+	Bullet.Ricochet = self.Primary.Ricochet
```
The `Bullet` struct doesn't have any properties named `Ricochet` or `MaxRicochet` and you don't seem to use them anywhere else in your code. Is this a mistake?

ELPatitoLoco replied:
Yes

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+function SWEP:PrimaryAttack()
+	local ply = self:GetOwner()
+
+	ply:LagCompensation(true)
```
`Weapon:FireBullets()` already uses `:LagCompensation()` internally.

ELPatitoLoco replied:
I used some old tutorials so could be that this is just obsolete now, thanks

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+	--self:DrawAmmo(1)
+	--self:DrawCrosshair(1)
+	self:SetNextPrimaryFire(CurTime() + self.Primary.Delay)
+	ply:LagCompensation(false)
```
Same as before: `Weapon:FireBullets()` already uses `:LagCompensation()` internally.

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+
+SWEP.LoadSound = "weapons/ammopickup.wav"
+
+SWEP.Primary.Recoil = 10
```
The `SWEP.Primary` table doesn't have any properties named `Recoil` and you don't seem to use them anywhere else in your code. Is this a mistake?

ELPatitoLoco replied:
I think the swep already has default revolver recoil so its a mistake but I didn't notice in the testing, anyways this is a early stage of the swep and it doesn't even have scope yet so its not big problem for now

ELPatitoLoco replied:
Came up with a reliable recoil

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.Primary.ClipSize = 10
+SWEP.Primary.DefaultClip = 10
+SWEP.Primary.Automatic = true
+SWEP.Primary.Ammo = "item_ammo_357_large"
```
`item_ammo_357_large` is the ammo entity, did you mean `357` the ammo type?

ELPatitoLoco replied:
yes, that could actually be why the swep actually fails completly at using ammo system

ELPatitoLoco replied:
Fixed thanks, anyways the ammo is disabled

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.RapidFireTimer = 0
+SWEP.RapidFireForce = 1000000
+
+SWEP.CannonForce = 50000000
```
The `SWEP` struct doesn't have a property named `CannonForce` and you don't seem to use them anywhere else in your code. Is this a mistake?

ELPatitoLoco replied:
Yes

ELPatitoLoco replied:
cannon force doesn't exist I think

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.Penetration = true
+SWEP.Ricochet = true
+SWEP.MaxRicochet = 100
+SWEP.Scoped = false
```
The `SWEP` struct doesn't have a property named `Scoped` and you don't seem to use them anywhere else in your code. Is this a mistake?

ELPatitoLoco replied:
The scope is a mess, I used old tutorials so i think the code they showed is obsolet/ no one has a tutorial or guide or info in using function:scope so that will be for a future fix

Kilometres commented on lua/weapons/E.R.C.S.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.Primary.Cone = 0.25
+SWEP.Primary.Delay = 1
+SWEP.Tracer = 4
+SWEP.Penetration = true
```
The `SWEP` struct doesn't have a property named `Penetration` and you don't seem to use them anywhere else in your code. Is this a mistake?

ELPatitoLoco replied:
Yes pretty much

ELPatitoLoco replied:
I actually use penetration

Kilometres reviewed (commented):
Overall, it was an amazing first SWEP (a lot fewer changes than my SWEP) but it seems like a few things may have been left in from tutorial code, or you may have conflated what addons do with what's on the Gmod wiki. Have a look through your code and double check that the properties you are setting actually do something.

Derpius commented:
FYI the capital letters in the filename may cause issues on Linux due to GMod assuming paths are case insensitive as on Windows (e.g., an addon folder called `starfallex` will load but `StarfallEx` wont).

Not sure if that affects more than addon folders though, but probably best to make it lowercase unless you've already verified that it works on Linux?

ELPatitoLoco commented:
> FYI the capital letters in the filename may cause issues on Linux due to GMod assuming paths are case insensitive as on Windows (e.g., an addon folder called `starfallex` will load but `StarfallEx` wont).
> 
> Not sure if that affects more than addon folders though, but probably best to make it lowercase unless you've already verified that it works on Linux?

I haven't tested on linux but I'll change file name, ig the weapon name can still be the same and won't cause conflicts

ELPatitoLoco committed:
Changing E.R.C.S.lua name to not create conflicts

ELPatitoLoco committed:
Rename E.R.C.S.lua to e.r.c.s.lua

Changing E.R.C.S.lua file name to lower case to not cause conflicts with linux

ELPatitoLoco committed:
Fixed lots of unused stuff and also added recoil

ELPatitoLoco committed:
Merge branch 'Swep_el_patito_E.R.C.S' of https://github.com/ELPatitoLoco/sweps.newswep into Swep_el_patito_E.R.C.S

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+	Bullet.AmmoType = self.Primary.Ammo
+	Bullet.Ricochet = self.Primary.Ricochet
+	Bullet.MaxRicochet = self.Primary.MaxRicochet
+	Bullet.Penetration = self.Primary.Penetration
```
`Ricochet`, `MaxRicochet` and `Penetration` are not properties of the `Bullet` struct and you do not use them anywhere else, could this be remnants of old code?

ELPatitoLoco replied:
I use penetration and it someless works I would make it stay for a while, the others are remanents of things that I need to remove or just for when I need to test something

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+	local rnda = self.Primary.Recoil * -1
+	local rndb = self.Primary.Recoil * math.random(1, 0)
+
+	ply:ViewPunch(Angle(rnda, rndb, rnda))
```
The variables `rnda`, `rndb` and `rnda` are not named very descriptively.

ELPatitoLoco replied:
Changed names with some ones more descriptive

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+	Bullet.Penetration = self.Primary.Penetration
+
+	local rnda = self.Primary.Recoil * -1
+	local rndb = self.Primary.Recoil * math.random(1, 0)
```
The parameters of `math.random(a, b)` are a range, so should be in ascending order. `math.random(1, 0)` always return `1` and never `0`.

ELPatitoLoco replied:
Also solved now its math.random(0,1)

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+
+SWEP.RapidFire = false
+SWEP.RapidFireTimer = 0
+SWEP.RapidFireForce = 1000000
```
`RapidFire`, `RapidFireTimer ` and `RapidFireForce ` are not properties of the `SWEP` struct and you do not use them anywhere else, could this be remnants of old code?

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+SWEP.Primary.Damage = 75
+SWEP.Primary.NumShots = 1
+SWEP.Primary.Spread = 0
+SWEP.Primary.Cone = 0.25
```
`Cone` is not a property of `SWEP.Primary` and you do not use it anywhere else, could this be remnants of old code?

ELPatitoLoco replied:
In a future when I add the scope and the secondary fire I will activate it but for now its just unused

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+
+local ShootSound = Sound("weapons/mortar/mortar_fire1.wav")
+
+SWEP.LoadSound = "weapons/ammopickup.wav"
```
`SWEP.LoadSound` is unused.

ELPatitoLoco replied:
Same as all, for the secondary fire

Kilometres commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,97 @@
+	self:FireBullets(Bullet)
+	self:ShootEffects()
+	self:EmitSound(ShootSound)
+	self.BaseClass.ShootEffects(self)
```
Is there a reason you aren't just using `SWEP:ShootEffects()`?

ELPatitoLoco replied:
Yes, It is more practical for me to do it this way, also I like to have the shootsound variable in hand, I know this may sound dumb but I like it this way

Kilometres replied:
In what way is it more practical?

Derpius replied:
@ELPatitoLoco there is no reason I can think of to do `self.BaseClass.ShootEffects(self)` over `self:ShootEffects()`, and your explanation doesn't really make sense

Kilometres reviewed (dismissed):
Comments:
- 80% of the changes are just removing unused properties, which doesn't change anything functionally but is inefficient and bad practice.
- 3 Variables are named obscurely and reduce code readability, especially for collaborators.
- Better practice alternative.

You're basically there, just a few tidy ups.

ELPatitoLoco committed:
Removed and commented a bunch of unused functions
and variables and changed random math having constant value

ELPatitoLoco committed:
Ultimate Commenting and fixing

Kilometres commented:
# QA Result: Failed ❌ 

## ❌ Blockers
- WorldModel does not load correctly. (There's a sneaky space in the model path)
![Invalid Model Path](https://github.com/TAServers/sweps/assets/39900063/b2c54689-2663-4228-89fe-3c2c9af54d39)
- Primary ammo isn't consumed. (but is setup as if it should be)
- Killfeed displays incorrectly.
![Killfeed](https://github.com/TAServers/sweps/assets/39900063/36aa1b70-d724-4535-bd58-83037f5327f4)

## ⚠️ Warnings
- Ammo not displayed. (Don't think this was intended)
- Yaw recoil being randomised discretely (either no yaw recoil or it recoils left) is a bit jarring, listen for user feedback. (Maybe use `math.Rand(-1, 1)` to get a random float, or use constant recoil, et cetera)
- No icon set in the spawnmenu.
![Spawnmenu Icon Blank](https://github.com/TAServers/sweps/assets/39900063/8fb44051-76c8-47d7-ac64-531ad04e5bcb)
- Spelling error in the description.
![Weapon Description](https://github.com/TAServers/sweps/assets/39900063/0f2c91bf-858f-4410-9f60-878747f10fe8)

ELPatitoLoco committed:
Merge branch 'TAServers:master' into Swep_el_patito_E.R.C.S

ELPatitoLoco committed:
Fixed a bunch of things, worldmodel not showing,
recoil, No Icon in spawnmenu, spelling errors, and some others

ELPatitoLoco committed:
Merge branch 'Swep_el_patito_E.R.C.S' of https://github.com/ELPatitoLoco/sweps.newswep into Swep_el_patito_E.R.C.S

ELPatitoLoco committed:
Removed unesecary parts, got some usefull comments
, got done some calibration to the bullets and
still having an issue with the kill feed

ELPatitoLoco committed:
Made the kill message functional by adding a
killicon.AddFont

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+SWEP.Author = "el_patito_loco"
+SWEP.PrintName = "e.r.c.s" --"Experimental Rotating Chamber Sniper"
```
Maybe make the full name the print name? (or put it in the description)

ELPatitoLoco replied:
Done, added the full name in description

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+	local RecoilMultiplier = self.Primary.Recoil * -1
+	local RecoilRandomMultiplier = self.Primary.Recoil * math.random(-1, 1)
+
+	ply:ViewPunch(Angle(RecoilMultiplier, RecoilRandomMultiplier, RecoilMultiplier))
```
The way this recoil is applied doesn't really make much sense:

- You're applying consistent recoil on pitch, but random on yaw
- You're applying the same amount of recoil on both pitch and yaw, but there should be minimal horizontal recoil just from the physics of conventional firearms
- You're applying recoil on the _roll_ axis, which doesn't seem right for similar reasons to above
- You're using `math.random` with two integers, which will return either -1, 0 or 1 (and not a float between -1 and 1). I think you want `math.Rand` here?

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+end
+
+function SWEP:SecondaryAttack() end
+--[[ All of this is seized until further improvment of my lua skills
+function SWEP:DoImpactEffect(GaussTracer,2)
+end
+function SWEP:Reload() end
+function SWEP:scope()
+	self:scope(1)
+end
+--]]
```
nit: Can this commented code be removed?

ELPatitoLoco replied:
Yes, I was a  thinking to do so but it was so far in code I didn't bothered to do it

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+end
+
+function SWEP:PrimaryAttack()
+	local ply = self:GetOwner()
```
nit: Can you use a more descriptive variable name here, e.g. `owner`

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+SWEP.Primary.NumShots = 10
+SWEP.Primary.Spread = 0
+SWEP.Primary.Delay = 1
+SWEP.MouseWasDown = true
```
This isn't a standard property on [the `SWEP` struct](https://wiki.facepunch.com/gmod/Structures/SWEP), and you don't seem to be using it in the code. Can it be removed?

ELPatitoLoco replied:
Yes, removed

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+
+--ammo
+
+SWEP.Primary.ClipSize = -1
+SWEP.Primary.DefaultClip = -1
+SWEP.Primary.Automatic = true
+SWEP.Primary.Ammo = "none"
```
A subjective choice, but should this weapon consume ammo?

ELPatitoLoco replied:
No, the weapon should not consume weapon, I was thinkins about adding a alt fire with ammo(even tho the thing you are talking about has nothing to do wtith it).

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+
+local ShootSound = Sound("weapons/mortar/mortar_fire1.wav")
+
+SWEP.FiresUnderwater = true
```
This property doesn't seem to exist on [the `SWEP` struct](https://wiki.facepunch.com/gmod/Structures/SWEP), has this been copied from a SWEP built on a weapon framework such as TFA, and can it be removed?

ELPatitoLoco replied:
Removed fireUnderwater, it was copy paste code

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,96 @@
+function SWEP:PrimaryAttack()
+	local ply = self:GetOwner()
+
+	local Bullet = {}
```
nit: There's inconsistent variable naming throughout this file. Could you stick to `camelCase` throughout?

ELPatitoLoco replied:
Named it owner (I guess you only refer to ply variable)

Derpius replied:
I was referring to `Bullet` and `ShootSound` (should be `bullet` and `shootSound` for consistency both within your script and existing code in the repository)

Derpius reviewed (commented):
In addition to these comments, will also need a re-QA from @Kilometres

ELPatitoLoco committed:
Removed a bunch of useless code and comments

Derpius commented on lua/weapons/e.r.c.s.lua:
```diff
@@ -0,0 +1,87 @@
+end
+
+if CLIENT then
+	killicon.AddAlias("e.r.c.s", "weapon_357") --Gives a killicon
+	language.Add("e.r.c.s", "E.R.C.S") -- Basically makes the killicon work
```
nit: These comments are redundant

Also `language.Add` doesn't affect the kill icon, it just sets the name of the weapon when displayed in the UI (e.g. in the kill feed)

Derpius reviewed (approved):
Happy to approve now, with a few nitpicks still open. Just needs a re-QA to merge

ELPatitoLoco committed:
Added more accurate comments

Kilometres commented:
Re-QA was successful with no blockers or warnings.
